examples/step-95/step-95.cc

```diff
@@ -0,0 +1,1715 @@
+  // function.
+  template <int dim>
+  class PoissonOperator
+  {
+    // We define 3 types frequently used throughout the tutorial keep the
+    // definition of the Number type at a central place.
+    using Number              = double;
+    using VectorizedArrayType = VectorizedArray<Number>;
+    using VectorType          = LinearAlgebra::distributed::Vector<Number>;
+
+    // We also define types for the matrix-free evaluation classes for cells and
+    // faces and for structured and unstructured quadrature to keep the
+    // definition of the template arguments for those at a central place.
+    using CellIntegrator =
+      FEEvaluation<dim, -1, 0, 1, Number, VectorizedArrayType>;
+    using FaceIntegrator =
+      FEFaceEvaluation<dim, -1, 0, 1, Number, VectorizedArrayType>;
+    using GenericCellIntegrator =
+      FEPointEvaluation<1, dim, dim, VectorizedArrayType>;
+    using GenericFaceIntegrator =
+      FEFacePointEvaluation<1, dim, dim, VectorizedArrayType>;
+
+    // And finally, we introduce a shortcut for the width of the detected SIMD
+    // vectorization.
+    static constexpr unsigned int n_lanes = VectorizedArrayType::size();
+
+  public:
+    // In the reinit() function the PoissonOperator receives the necessary
+    // information for matrix-free evaluation, i.e. the MatrixFree object for
+    // inside cells and the NonMatching::MappingInfo objects for intersected
+    // cells.
+    void
+    reinit(const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free_in,
+           const NonMatching::MappingInfo<dim, dim, VectorizedArrayType>
+             *mapping_info_cell_in,
+           const NonMatching::MappingInfo<dim, dim, VectorizedArrayType>
+             *mapping_info_surface_in,
+           const NonMatching::MappingInfo<dim, dim, VectorizedArrayType>
+                             *mapping_info_faces_in,
+           const bool         is_dg_in,
+           const unsigned int degree)
+    {
+      matrix_free          = &matrix_free_in;
+      mapping_info_cell    = mapping_info_cell_in;
+      mapping_info_surface = mapping_info_surface_in;
+      mapping_info_faces   = mapping_info_faces_in;
+      is_dg                = is_dg_in;
+
+      fe_dgq = std::make_unique<FE_DGQ<dim>>(degree);
+
+      evaluator_cell =
+        std::make_unique<GenericCellIntegrator>(*mapping_info_cell, *fe_dgq);
+      evaluator_surface =
+        std::make_unique<GenericCellIntegrator>(*mapping_info_surface, *fe_dgq);
+      if (is_dg)
+        {
+          evaluator_face_m =
+            std::make_unique<GenericFaceIntegrator>(*mapping_info_faces,
+                                                    *fe_dgq,
+                                                    true);
+          evaluator_face_p =
+            std::make_unique<GenericFaceIntegrator>(*mapping_info_faces,
+                                                    *fe_dgq,
+                                                    false);
+        }
```

Comment:
Some context for #18356. This is the situation that PR tries to avoid. Using a `FE_DGQ` element for the constructor currently tells `FEPointEvaluation` to use lexicographic numbering internally.

The problem is the mismatch of DoF numbering of `DoFCellAccessor::get_dof_values` (hierarchic) and `FEEvaluation::read_dof_values` (lexicographic) for `FE_Q` elements. As proposed in #18356, I think a constructor argument and an adjustment of the current documentation is the way to go. Then this "hack" can be eliminated here (and in my user code).